### ISSUE TITLE
Stop publishing rxjs-es. Move ES6 build to output .mjs files alongside the CJS .js files.

### ISSUE THREAD

**Issue by blesh** [ReactiveX/rxjs#1671]
In speaking with @robwormald the Angular folks are forced to swap between `rxjs` and `rxjs-es` during their build to enable proper tree-shaking. This is probably less than ideal. If all of the es6 files were under the same directories as `.mjs` files, then it would be as easy as configuring the bundler to look at that file extension to use ES6 modules.

This would mean that the `rxjs-es` build would likely not have a use anymore.

cc/ @IgorMinar @jeffbcross @bradlygreen

**Comment by jayphelps** [comment]
I'm not big on jumping on the `.mjs` bandwagon so dang soon, particularly since it is so controversial..it's far from decided.

That said, if we did publish `.mjs` files alone size the `.js` CJS files, it wouldn't _hurt_ anyone but ourselves down the road if the `.mjs` thing dies but people are still depending on `rxjs` supporting it. So if we're OK with committing to support it even if `.mjs` dies (at least for a transitional period), then I'm not otherwise against it.

Also--I do not think we should stop publishing `rxjs-es` regardless because there will definitely be people who don't subscribe to the `.mjs` pitch but still want to consume the raw ES6 module version via standard `.js` extensions.

**Comment by robwormald** [comment]
should be clear - i'm not advocating at all the `.mjs` bandwagon - in fact we're currently on the `jsnext:main` bandwagon (via rollup) - which ideally has es6 sources distributed in a subfolder of the cjs distro and a field in the package.json pointing at it's entry point.

Sigh.

**Comment by jayphelps** [comment]
@robwormald Great clarification, I didn't realize that. oy vey module systems...

I'm way more OK with including the source as ES6 modules w/ `.js` extensions inside the normal `rxjs` CJS package, but inside a dir like `es6/*.js` and then pointing `jsnext:main` to it.

**Comment by jayphelps** [comment]
Just to make things more interesting, TypeScript does not currently support targeting ES5 but keeping ES6 module syntax. e.g. `tsc --target es5 --module es6` doesn't work.

Open issue: https://github.com/Microsoft/TypeScript/issues/6319

Sooooo...we could output entirely ES6 code, but then you _must_ transpile all ES6 syntax, not just the module syntax. Is that OK?

**Comment by jayphelps** [comment]
hahaha actually, we could build `esm/*.js` files by building rxjs in TS -> ES6 then pipe that into babel to output ES5 w/ ES6 module syntax...gawd help us.

**Comment by robwormald** [comment]
we (currently) figure that just outputting full on ES6 code there is probably fine for the moment. Note that latest versions of TS allow .js input with `--allowJs` so you can use a single tool across the chain - i'm currently doing TS -> ES6 -> rollup -> ES5 via tsc.

**Comment by robwormald** [comment]
We're doing the `esm` subdir thing. Makes as much (or as little...) sense as anything right now :shipit:

**Comment by blesh** [comment]
So, @robwormald and @jayphelps ... are we currently, then, proposing to moving to build everything out into one package? 

Basically, what we currently publish to `rxjs` (which is the CJS files), but also with an `esm` folder? (and perhaps a `ts` folder, I guess?)

In turn, we'd have one UMD output, and would no longer produce a special AMD bulid?

**Comment by robwormald** [comment]
that all seems reasonable to me. we don't need a `ts` subdirectory, as long as the `d.ts` files are distributed alongside. so it should look like
```

/Observable.js //commonjs
/Observable.d.ts
/esm/Observable.js //es6
/esm/Observable.d.ts
```

**Comment by kylecordes** [comment]
Over on some much less complex libraries than RxJS, we have been doing the same thing as @robwormald  explained above, with excellent results. It "just works" and can be consumed without special tooling effort both as commonJS ES5 and ES6 with ES6 modules.  Lacking any better plan (and with similar hesitation on the MJS thing) I think this would be an excellent step forward for RxJS.

**Comment by IgorMinar** [comment]
+1

there are no great options out there right now, but the `jsnext:main` as
used by angular is the thing that works with some caveats. one of them
should be researched by rxjs - deep imports will break in the esm mode.
This means that we need a story for `add/operator/foo` imports because
these won't work out of the box.

keep in mind that there are two kinds of path resolutions going on:

1/ typescript does a resolution when looking for type declarations stored
in d.ts files - typescript currently doesn't consider jsnext:main and will
grab cjs d.ts files. This is likely ok because the d.ts files are currently
the same for cjs and esm. For this mode the deep imports are ok.
2/ bundler (rollup, webpack, system builder) resolves all import paths and
all of the support `jsnext:main` entry point. For this mode deep imports
won't work out of the box and special care needs to be taken to make them
work.

we (angular) opted to not support deep imports except for a few very
special cases (e.g. cases when always defaulting to cjs is ok - like when
providing e2e test utilities).

**Comment by blesh** [comment]
@IgorMinar, interesting. I'd there a write up as to why deep imports don't work with bundlers?

**Comment by robwormald** [comment]
The tldr is that while its easy to remap an entry point file via something like `jsnext:main`, its not quite as trivial to remap a whole folder, and you end up having to do more configuration in your bundling tools. Not a huge deal but it adds some mental overhead.

**Comment by blesh** [comment]
@robwormald @IgorMinar so what's actionable out of this issue?

Seems like:

1. Change ES6 build to build to an `es` or `esm/` folder. (Probably just targeting ES5 features with ES6 modules if possible)
2. Add `jsnext:main` to `package.json`

Is that it?

Related: I also want to stop building the AMD stuff in favor of UMD, and rename the UMD output file to just `Rx.js` rather than `Rx.umd.js`

**Comment by IgorMinar** [comment]
3. We need to figure out how to make adding operators work.

**Comment by kylecordes** [comment]
As an expedient way to get the current mechanism easily importable via ES6, one easyish answer is to mechanically create a top level module in the RxJS package which exports a bunch of separate add functions, one for each operator. I think there is a way to shape this file correctly that the packaging mechanism only picks up the specific things you want. So instead of this:

```
import "rxjs/add/operator/map";
```

You do this:

```
import {addMap} from "rxjs/addOperators";
```

?

**Comment by IgorMinar** [comment]
that wouldn't work for importing type information for that operator.

**Comment by blesh** [comment]
@robwormald could we codegen remappings for people at build time and distribute it with the package?

**Comment by kylecordes** [comment]
> importing type information

Ah, I had forgotten the very clever type set up. Okay, here are more ideas:

```
import "rxjs/add-operator-map";
```

It's not all that bad to look at, and it fits in the ability to have just one "/" in the path.

```
import "@rxjs/add-operator/map";
```

By this scheme we get two "/" and still have support for node resolution with esnext.  Obviously this implies shipping rxjs as a set of packages, like Angular itself.

Either of the these things could be generated rather than and coded.

This latter model is a bit of a blast the past, it wasn't that many months ago when this was all in scoped packages. I don't know the motivation for moving it back out to a top-level package, but now that Angular itself is in the scope package certainly everyone is going to find them acceptably familiar and all the tooling is going to have to work well enough with them anyway.

**Comment by jayphelps** [comment]
btw TypeScript is very shortly going to support `tsc --target es5 --module es6` (i.e. transpile everything except output ES modules)  https://github.com/Microsoft/TypeScript/pull/9042

**Comment by blesh** [comment]
Revisiting this... @IgorMinar, can you clarify this statement?

>  For this mode deep imports won't work out of the box and special care needs to be taken to make them work.

Why will deep imports not work? if `import 'rxjs/Observable'` looks at `/esm/Observable.js` why would `import 'rxjs/add/operator/map'` not look at `/esm/add/operator/map.js`?

**Comment by robwormald** [comment]
> Why will deep imports not work? if import 'rxjs/Observable' looks at /esm/Observable.js why would import 'rxjs/add/operator/map' not look at /esm/add/operator/map.js

Because, apparently, that's just the way it is, see: https://github.com/rollup/rollup-plugin-node-resolve/issues/44

What we need is agreement (LO-freaking-L) on a `jsnext:dir`

**Comment by robwormald** [comment]
FYI, we've implemented this structure in Angular2:

index.js (ES5 + ES Modules)
bundles/core.umd.js (UMD + ES5)

package.json:
- main : bundles/core.js
- module: index.js

So far, so good. This allows commonJS users to simply `require('rxjs')`, and browser users to deep import w/o worrying about paths.

**Comment by danbucholtz** [comment]
The .js files in `rxjs-es` contain es2015 code (let, const, arrow functions, etc). The es2015 module declarations are great for tree shaking but the code needs to be transpiled to es5 to be used by most browsers/apps.

Our use case was tree shaking out unused `rxjs` stuff from an Angular 2 library (Ionic Framework).

We are using the `module` field in our bundler to use the `es2015` modules for Angular, and we would love to see `rxjs` do the same. 😃 

Thanks so much,
Dan

**Comment by blesh** [comment]
Okay, this is really the last issue blocking release, AFAICT... I feel we have a few options:

1. Stop publishing the ES Module build entirely prior to release
  - Remove this build before 5.0 release
  - bikeshed what we want to do for ES2015 module support, but still give everyone the .ts files with which they can build whatever processes they want
  - Once we decide on a strategy, we implement it and it's only a _minor_ version change (v5.1 for example)
2. Publish .mjs files on `rxjs` package and cross our fingers a breaking change isn't required later.
3. Continue publishing `rxjs-es` and cross our fingers and cross our fingers a breaking change isn't required later.

Both 2 and 3 still require the change that we should be outputting ES5 code with ES2015 modules.

Looking at the npm stats, `rxjs-es` is not even close to as broadly used as `rxjs`. We're talking a couple of orders of magnitude. It's clearly not a high value for our users, but seems to be of *some* value.

**Comment by blesh** [comment]
@robwormald wouldn't [the plan you outlined above](https://github.com/ReactiveX/rxjs/issues/1671#issuecomment-225324611) prevent "tree-shaking" and/or encourage usage that would forgo the advantages of modularity in RxJS 5?

**Comment by danbucholtz** [comment]
The Ionic team is following suit with what Angular is doing and encouraging others to do the same. 

I am producing documentation for library makers on how to get their lib into that state and what sort of build step changes they need to make.

@blesh, no it would not because bundlers will read the `module` option before the `main` option, and `module` points to esm code.

Thanks,
Dan

**Comment by jayphelps** [comment]
After discussing this further, we're not entirely sure of the actual benefits of tree shaking for RxJS itself. This is primary because RxJS is almost entirely operator based. So let me elaborate.

#### If you import everything a la `import Rx from 'rxjs';`

All the operators are added to either `Observable.XXX` or `Observable.prototype.XXX`. So right away all the operators are "unshakable" because they are found on objects and you cannot use static analysis to _guarantee_ they are not used somewhere. You could place restrictions on certain JS features to make that guarantee, but to my knowledge Rollup, Webpack, etc do not attempt or plan to attempt anything like this--instead shaking modules by detecting used export bindings along with verifying there are no internal references to that binding. In this case, there are indeed internal references to the operator code, so even though they are exported they cannot be tossed because they could in fact be used.

```js
import Rx from 'rxjs';
const operatorChoice = prompt('which operator?');
Observable[operatorChoice]();
```

It's similar to how in webpack you can't do `require(prompt('which module?'))` for much more obvious reasons.

#### If you import operators individually to prototype `import 'rxjs/add/operator/map';`

There is nothing to shake. You are only importing operators you intend use, which they will be added to the Observable prototype as a side effect. Even if you didn't end up using the ones you import, again static analysis cannot confirm this with certainty.

#### If you import operators individually and apply them as needed `import { map } 'rxjs/operator/map';`

Here you could in theory do some shaking in the sense that if someone imports an operator but in fact never references it in that module at all. I'm not sure if Rollup/webpack do this. My guess would be no because it provides little practical value. If you're not using something, don't import it. Very often people have lint rules that detect unused references/imports.

#### Things other than operators

We next thought about things that aren't operators: the different Subjects, Schedulers, etc.

Nearly all of them are imported and consumed by _some_ operator. e.g. `publishBehavior = BehaviorSubject + ConnectableObservable`. Though certainly there will be some comparatively tiny things that could be shaken off like `TestScheduler`. An argument could be made that small amount of code is worth it, so I'm not outright saying it's not.

***

In summary, we're skeptical of the practical benefits. Can anyone see something we missed?  Maybe we're missing the point. Is there something other than tree shaking that this enables?

**Comment by thelgevold** [comment]
I randomly saw this issue and got curious...

In the Angular 2 docs we use the `commonjs` plugin to convert rx-js commonjs to esm during bundling.
https://github.com/rollup/rollup-plugin-commonjs
Is there a practical difference/benefit between doing that and targeting released esm source like es-rxjs directly? 

If there is no practical difference, isn't this easier than publishing two versions of the source? 

I remember there was an extra step integrating the es-rxjs code with rollup but not too bad at the end of the day though.

```
export default {
  entry: '../../toh-6/ts/app/main-aot.js',
  dest: 'dist/build.js', // output a single application bundle
  sourceMap: true,
  sourceMapFile: 'dist/build.js.map',
  format: 'iife',
  plugins: [
      nodeResolve({jsnext: true, module: true}),
      commonjs({
        include: ['node_modules/rxjs/**']// convert the rx-js commonjs code here
      }),
      uglify()
  ]
}
```

**Comment by Rich-Harris** [comment]
@jayphelps I think your assessment is correct (at least based on my relatively slim knowledge of Rx) that tree-shaking will provide minimal benefits as things stand. 

> You could place restrictions on certain JS features to make that guarantee, but to my knowledge Rollup, Webpack, etc do not attempt or plan to attempt anything like this

We do have vague plans to improve the tree-shaking to the level that Rollup is able to remove unused class methods etc, but it's all a bit blue sky at the moment with no actual working code. If it does see the light of day, it won't be for a while.

My own inclination if `rxjs-es` is being killed off would be to ship deep imports in an `esm` subfolder (ES5 + modules), so that people still have a route to use ESM and deep imports without inconveniencing the majority – they might be a couple orders of magnitude smaller right now but I suspect once Webpack 2 ships there's going to be a surge of interest in this sort of thing as people come to understand the benefits. I'd definitely steer clear of `.mjs`!

**Comment by jayphelps** [comment]
@Rich-Harris thanks for the insight! 👍 

***

@thelgevold I don't quite follow what you're suggesting. We would need to publish two versions of the source, both in ES5 but differing module formats CJS/ES. I think you are prolly referring to _npm packages_ e.g. `rxjs` vs. `rxjs-es` which is a valid point. If the most popular ES module bundlers use the `module` field we could indeed ship both versions within the same package. I think this is a likely outcome, but it's not without negatives. Somewhere around doubling the size of our npm package isn't trivial, particularly because some notable Electron users of Rx have already expressed issue with our existing package sizes; Slack being one, Nuclide the other. Doubling it obviously is a move in the opposite direction.

We may still do it, just wanted to bring up some concerns.

**Comment by thelgevold** [comment]
@jayphelps I was thinking that the commonjs -> esm conversion in the bundler (via plugin) would remove the need for publishing esm at all. Wouldn't it be enough to only publish the commonjs source like in the current rxjs npm package? 

I might be missing something though :-)

**Comment by jayphelps** [comment]
@thelgevold ESM -> CJS -> ESM is a lossy conversion [in theory](https://github.com/rollup/rollup-plugin-commonjs#custom-named-exports); which is why true tree shaking is only done with ESM and that plugin provides ways for you to "tell it" how to handle ambiguous cases. In practice it might not matter for Rx's usage of exports, I'm not sure. Either way I'm inclined to distribute lossless (true) ESM builds so that at the very least people stop bugging us about it 😆. It is indeed the latest fad.

**Comment by thelgevold** [comment]
@jayphelps That makes sense. Thanks for the explenation

**Comment by jeffbcross** [comment]
> Somewhere around doubling the size of our npm package isn't trivial, particularly because some notable Electron users of Rx have already expressed issue with our existing package sizes;

Why, are they shipping the full contents of `node_modules` folder with their applications?

**Comment by blesh** [comment]
At this point, I'm leaning in favor of just ceasing to publish `rxjs-es` ahead of the v5.0 release. Given that:

1. Rollup can only tree-shake about 1% of RxJS 5, so it's not really a great way to use this library.
2. We can bike shed how we want to publish es2015 modules all we want and then add it later without it being a breaking change.
3. Using rxjs-es is confusing and is of limited popularity anyhow.

**Comment by dbazile** [comment]
Would someone be able to regenerate the docs for http://reactivex.io/rxjs/manual/installation.html#es6-via-npm?  As of 2017-08-22, it still recommends `npm install rxjs-es`.

I just installed rxjs-es (thinking it was still in favor) and stumbled upon this thread at #2218 after TypeScript raised a fit about file name casing. 😅 

![screen shot 2017-08-22 at 9 11 09 pm](https://user-images.githubusercontent.com/1542531/29594474-306aadf4-877f-11e7-927d-8e85b7ad20db.png)
